### PR TITLE
Module loader

### PR DESCRIPTION
**Pull request by florian-vuillemot**
Module loader for windows and linux

### CODE REVIEW

File: src/moduleLoader/DynamicLink.hpp [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    }
+  }
+
+  std::shared_ptr<T> get(std::string const &moduleName) {
+    for (auto &loader: loaders_){
+      if (loader.getFilename() == moduleName){
+	return loader.create();
+      }
+    }
+    return std::shared_ptr<T>();
```
Comment (polosskevin) [inline src/moduleLoader/DynamicLink.hpp]: you can return nullptr here

File: src/moduleLoader/ModuleLoader.hpp [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+ * - "create_object" that return a instance of object.
+ * - "destroy_object" that call the destructor of object.
+ */
+template<class T>
+class ModuleLoader {
+public:
+    /**
+     * Mandatory function that return a T*.
+     */
+    std::string const externCreateObject = "create_object";
```
Comment (polosskevin) [inline src/moduleLoader/ModuleLoader.hpp]: can you make those configurable via setter or constructor ?

File: src/moduleLoader/ModuleLoader.hpp [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+   
+#ifdef __linux__
+    auto cr = (T *(*)()) dlsym(sdl_library, externCreateObject.c_str());
+    auto dr = (void (*)(T *)) dlsym(sdl_library, externDestructorObject.c_str());
+#elif _WIN32
+    auto cr = GetProcAddress(sdl_library, externCreateObject.c_str());
+    auto dr = (void (*)(T *))GetProcAddress(sdl_library, externDestructorObject.c_str());
+#endif
+    
+    return std::shared_ptr<T>(cr(), [dr](void *ptr){
```
Comment (polosskevin) [inline src/moduleLoader/ModuleLoader.hpp]: why not use T * here

File: src/moduleLoader/ModuleLoader.hpp (hunk 2 above)
Comment (florian-vuillemot) [reply src/moduleLoader/ModuleLoader.hpp]: I could do this, but I don't think is a good idea. I think all module should have this C function.
Should I change there ?

File: src/moduleLoader/ModuleLoader.hpp (hunk 3 above)
Comment (florian-vuillemot) [reply src/moduleLoader/ModuleLoader.hpp]: You a right

File: src/moduleLoader/ModuleLoader.hpp (hunk 2 above)
Comment (polosskevin) [reply src/moduleLoader/ModuleLoader.hpp]: no it's good

File: src/moduleLoader/DynamicLink.hpp (hunk 1 above)
Comment (florian-vuillemot) [reply src/moduleLoader/DynamicLink.hpp]: Great, I update it.

**Pull request merged**